File: src/normalize.py
"""Deterministic parsing of the raw strings transcribed by the model.

This is where every number, unit and reference range is *computed* — the LLM
never does arithmetic. Kept small and pure so it can be unit-tested against the
tricky Czech cases (decimal commas, thousands separators, censored values,
open-ended ranges). A misread decimal is the cardinal bug this module guards.
"""
from __future__ import annotations

import re
from typing import Optional

from .models import Flag, Measurement

# Characters -----------------------------------------------------------------
_MICRO_VARIANTS = {"μ": "µ"}          # Greek mu → micro sign
_DASHES = {"‒", "–", "—", "−"}  # figure/en/em dash, minus
_THIN_SPACES = {" ", " ", " ", " "}  # NBSP, narrow NBSP, thin, figure

_NUMBER_CORE = re.compile(r"^[+]?\d[\d\s.,]*$")
_HAS_LETTER = re.compile(r"[^\W\d_]", re.UNICODE)  # any unicode letter


# --- numbers ----------------------------------------------------------------
def parse_czech_number(raw: Optional[str]) -> Optional[float]:
    """Parse a Czech-formatted number: comma decimal, space thousands.

    "5,4" -> 5.4 ; "1 234,5" -> 1234.5 ; "0,443" -> 0.443 ; "114" -> 114.0.
    Returns None if the string is not a plain number.
    """
    if raw is None:
        return None
    s = raw.strip()
    if not s:
        return None
    for bad, good in _MICRO_VARIANTS.items():
        s = s.replace(bad, good)
    for sp in _THIN_SPACES:
        s = s.replace(sp, "")
    s = re.sub(r"\s+", "", s)
    if not _NUMBER_CORE.match(s):
        return None
    # Both separators present: "." is the thousands grouping, drop it.
    if "," in s and "." in s:
        s = s.replace(".", "")
    s = s.replace(",", ".")
    try:
        return float(s)
    except ValueError:
        return None


def parse_value(value_raw: Optional[str]) -> Optional[float]:
    """Numeric value of a result cell, or None for censored/qualitative cells.

    Censored ("<1,0", ">10") and qualitative ("neprovedeno", "negativní")
    values deliberately return None — we never invent a number for a trend.
    """
    if value_raw is None:
        return None
    s = value_raw.strip()
    # Strip lab out-of-range markers ("!", "*") — decoration, not the number.
    s = s.replace("!", "").replace("*", "").strip()
    if not s:
        return None
    if "<" in s or ">" in s:
        return None
    if _HAS_LETTER.search(s):
        return None
    return parse_czech_number(s)


# --- units ------------------------------------------------------------------
def canonicalize_unit(unit_raw: Optional[str]) -> Optional[str]:
    """Fold cosmetic unit variants to one form so the same analyte lines up.

    Unifies micro-sign codepoints, the "10^9" vs "10˄9" exponent glyphs, and
    the litre-case ("/L" vs "/l"). Dimensionless markers ("-", "") -> "".
    """
    if unit_raw is None:
        return None
    s = unit_raw.strip()
    if s in {"", "-", "–", "—"}:
        return ""
    for bad, good in _MICRO_VARIANTS.items():
        s = s.replace(bad, good)
    s = s.replace("˄", "^")            # modifier caret ˄ → ^
    s = re.sub(r"(?i)/l\b", "/l", s)         # litre symbol case
    s = re.sub(r"\s+", " ", s).strip()
    return s


# --- reference ranges -------------------------------------------------------
def parse_range(ref_raw: Optional[str]) -> tuple[Optional[float], Optional[float], Optional[str]]:
    """Parse a printed reference range into (low, high, text).

    Handles "4,11-5,60", "62,00 - 110", "< 5,00", "> 0,5" and non-numeric
    ranges ("negativní"). The raw string is always preserved by the caller, so
    an unparseable range degrades to text rather than being dropped.
    """
    if ref_raw is None:
        return None, None, None
    s = ref_raw.strip().strip("()").strip()
    if not s:
        return None, None, None
    for d in _DASHES:
        s = s.replace(d, "-")

    if s.startswith("<"):
        high = parse_czech_number(s[1:])
        return (None, high, None) if high is not None else (None, None, s)
    if s.startswith(">"):
        low = parse_czech_number(s[1:])
        return (low, None, None) if low is not None else (None, None, s)

    # a - b  (both bounds). Split on the first hyphen that sits between digits.
    m = re.match(r"^\s*([0-9][0-9\s.,]*?)\s*-\s*([0-9][0-9\s.,]*)\s*$", s)
    if m:
        low = parse_czech_number(m.group(1))
        high = parse_czech_number(m.group(2))
        if low is not None and high is not None:
            return low, high, None

    # Non-numeric (e.g. "negativní") — keep as descriptive text.
    if _HAS_LETTER.search(s):
        return None, None, s
    return None, None, s


# --- flag -------------------------------------------------------------------
def compute_flag(
    value: Optional[float],
    low: Optional[float],
    high: Optional[float],
) -> Flag:
    """Recompute normal/low/high/unknown from value vs. parsed range.

    We never trust the lab's printed flag glyph — this is lab-independent.
    """
    if value is None:
        return "unknown"
    if low is None and high is None:
        return "unknown"
    if low is not None and value < low:
        return "low"
    if high is not None and value > high:
        return "high"
    return "normal"


# --- measurement ------------------------------------------------------------
def normalize_measurement(m: Measurement) -> Measurement:
    """Fill the derived numeric fields on a Measurement in place."""
    m.value = parse_value(m.value_raw)
    m.unit = canonicalize_unit(m.unit_raw)
    m.ref_range_low, m.ref_range_high, m.ref_range_text = parse_range(m.ref_range_raw)
    m.flag = compute_flag(m.value, m.ref_range_low, m.ref_range_high)

    # QA: downgrade confidence when a numeric-looking value failed to parse or a
    # numeric-looking range failed to parse — these are exactly the rows the
    # verify UI must surface.
    if m.value is None and m.value_raw.strip() and not _HAS_LETTER.search(m.value_raw) \
            and "<" not in m.value_raw and ">" not in m.value_raw:
        m.confidence = "low"
    return m

File: web/src/lib/normalize.ts
/**
 * Deterministic parsing of the raw strings a model transcribed.
 *
 * Ported from src/normalize.py, kept pure and side-effect free so the same
 * assertions the Python tests make can be made here. This is where every
 * number, unit and reference range is *computed* — the model never does
 * arithmetic. A misread decimal is the cardinal bug this module guards.
 */
import type { Confidence, Flag, Measurement } from "./models";

const MICRO_VARIANTS: Array<[string, string]> = [["μ", "µ"]]; // Greek mu → micro sign
const DASHES = ["‒", "–", "—", "−"]; // figure/en/em dash, minus
const THIN_SPACES = [" ", " ", " ", " "]; // NBSP, narrow NBSP, thin, figure

const NUMBER_CORE = /^\+?\d[\d\s.,]*$/;
const HAS_LETTER = /\p{L}/u;

function applyMicro(s: string): string {
  for (const [bad, good] of MICRO_VARIANTS) s = s.split(bad).join(good);
  return s;
}

/**
 * Parse a Czech-formatted number: comma decimal, space thousands.
 * "5,4" → 5.4 ; "1 234,5" → 1234.5 ; "114" → 114. Non-numbers → null.
 */
export function parseCzechNumber(raw: string | null | undefined): number | null {
  if (raw === null || raw === undefined) return null;
  let s = raw.trim();
  if (!s) return null;
  s = applyMicro(s);
  for (const sp of THIN_SPACES) s = s.split(sp).join("");
  s = s.replace(/\s+/g, "");
  if (!NUMBER_CORE.test(s)) return null;
  // Both separators present: "." is the thousands grouping, drop it.
  if (s.includes(",") && s.includes(".")) s = s.split(".").join("");
  s = s.split(",").join(".");
  const n = Number(s);
  return Number.isFinite(n) ? n : null;
}

/**
 * Numeric value of a result cell, or null for censored/qualitative cells.
 * Censored ("<1,0") and qualitative ("neprovedeno") deliberately return null —
 * we never invent a number for a trend.
 */
export function parseValue(valueRaw: string | null | undefined): number | null {
  if (valueRaw === null || valueRaw === undefined) return null;
  // Strip lab out-of-range markers ("!", "*") — decoration, not the number.
  const s = valueRaw.trim().split("!").join("").split("*").join("").trim();
  if (!s) return null;
  if (s.includes("<") || s.includes(">")) return null;
  if (HAS_LETTER.test(s)) return null;
  return parseCzechNumber(s);
}

/**
 * Fold cosmetic unit variants to one form so the same analyte lines up across
 * labs. Dimensionless markers ("-", "") → "".
 */
export function canonicalizeUnit(unitRaw: string | null | undefined): string | null {
  if (unitRaw === null || unitRaw === undefined) return null;
  let s = unitRaw.trim();
  if (s === "" || s === "-" || s === "–" || s === "—") return "";
  s = applyMicro(s);
  s = s.split("˄").join("^"); // modifier caret ˄ → ^
  s = s.replace(/\/l\b/gi, "/l"); // litre symbol case
  s = s.replace(/\s+/g, " ").trim();
  return s;
}

export interface ParsedRange {
  low: number | null;
  high: number | null;
  text: string | null;
}

/**
 * Parse a printed reference range into low/high/text. Handles "4,11-5,60",
 * "< 5,00", "> 0,5" and non-numeric ranges ("negativní"). An unparseable range
 * degrades to text rather than being dropped.
 */
export function parseRange(refRaw: string | null | undefined): ParsedRange {
  const none: ParsedRange = { low: null, high: null, text: null };
  if (refRaw === null || refRaw === undefined) return none;
  let s = refRaw.trim().replace(/^[()]+|[()]+$/g, "").trim();
  if (!s) return none;
  for (const d of DASHES) s = s.split(d).join("-");

  if (s.startsWith("<")) {
    const high = parseCzechNumber(s.slice(1));
    return high !== null ? { low: null, high, text: null } : { low: null, high: null, text: s };
  }
  if (s.startsWith(">")) {
    const low = parseCzechNumber(s.slice(1));
    return low !== null ? { low, high: null, text: null } : { low: null, high: null, text: s };
  }

  // a - b (both bounds): split on the first hyphen sitting between digits.
  const m = /^\s*([0-9][0-9\s.,]*?)\s*-\s*([0-9][0-9\s.,]*)\s*$/.exec(s);
  if (m) {
    const low = parseCzechNumber(m[1]);
    const high = parseCzechNumber(m[2]);
    if (low !== null && high !== null) return { low, high, text: null };
  }

  // Non-numeric (e.g. "negativní") — keep as descriptive text.
  return { low: null, high: null, text: s };
}

/**
 * Recompute normal/low/high/unknown from value vs. parsed range.
 * We never trust the lab's printed flag glyph — this is lab-independent.
 */
export function computeFlag(
  value: number | null,
  low: number | null,
  high: number | null,
): Flag {
  if (value === null) return "unknown";
  if (low === null && high === null) return "unknown";
  if (low !== null && value < low) return "low";
  if (high !== null && value > high) return "high";
  return "normal";
}

/**
 * Fill the derived numeric fields on a measurement, returning a new object.
 *
 * Returning a copy rather than mutating (as the Python does) matters in the
 * UI: correcting a value in the verify tab re-runs this and React needs a new
 * reference to notice.
 */
export function normalizeMeasurement(m: Measurement): Measurement {
  const value = parseValue(m.valueRaw);
  const unit = canonicalizeUnit(m.unitRaw);
  const { low, high, text } = parseRange(m.refRangeRaw);
  const flag = computeFlag(value, low, high);

  // QA: a numeric-looking value that failed to parse is exactly the row the
  // verify UI must surface.
  let confidence: Confidence = m.confidence;
  const vr = m.valueRaw.trim();
  if (value === null && vr && !HAS_LETTER.test(vr) && !vr.includes("<") && !vr.includes(">")) {
    confidence = "low";
  }

  return {
    ...m,
    value,
    unit,
    refRangeLow: low,
    refRangeHigh: high,
    refRangeText: text,
    flag,
    confidence,
  };
}

/** A row needing a human: low confidence, model disagreement, or unparseable. */
export function isFlagged(m: Measurement): boolean {
  return (
    m.confidence === "low" ||
    m.disagreement !== null ||
    (m.value === null && m.valueRaw.trim() !== "" && !HAS_LETTER.test(m.valueRaw))
  );
}
